### ISSUE TITLE
Validation of action output fails

### ISSUE THREAD

**Issue by milanlenco** [CESNET/libyang#153]
Validation of type RPCREPLY fails for any non-top-level action due to incorrect pre-processing of options. This patch fixes the issue:

```
diff --git a/src/tree_data.c b/src/tree_data.c
index b9e82ee..2cadafb 100644
--- a/src/tree_data.c
+++ b/src/tree_data.c
@@ -3612,7 +3612,7 @@ lyd_validate(struct lyd_node **node, int options, void *var_arg)
         }
     }
 
-    if ((options & LYD_OPT_RPC) && *node && ((*node)->schema->nodetype != LYS_RPC)) {
+    if (((options & LYD_OPT_RPC) || (options & LYD_OPT_RPCREPLY)) && *node && ((*node)->schema->nodetype != LYS_RPC)) {
         options |= LYD_OPT_ACT_NOTIF;
     }
     if ((options & LYD_OPT_NOTIF) && *node && ((*node)->schema->nodetype != LYS_NOTIF)) {
@@ -5581,7 +5581,7 @@ lyd_defaults_add_unres(struct lyd_node **root, int options, struct ly_ctx *ctx,
             LOGERR(LY_EINVAL, "Not valid RPC/action data.");
             return EXIT_FAILURE;
         }
-        if ((options & LYD_OPT_RPCREPLY) && !((*root)->schema->nodetype & (LYS_RPC | LYS_ACTION))) {
+        if ((options & LYD_OPT_RPCREPLY) && !act_notif && !((*root)->schema->nodetype & (LYS_RPC | LYS_ACTION))) {
             LOGERR(LY_EINVAL, "Not valid reply data.");
             return EXIT_FAILURE;
         }
```

**Comment by michalvasko** [comment]
Hi Milan,
based on the YANG 1.1 RFC [example](https://tools.ietf.org/html/rfc7950#section-7.15.3), actions are invoked with all their parents, but replies are identical to standard RPC replies, so I believe the code is correct.

Regards,
Michal

**Comment by milanlenco** [comment]
Hi Michal,
ah, ok, but now I have difficulties to get the validation of action output to pass through the initial stages of the validation. As I construct the action output (using lyd_new_path and co.) I get a subtree in which the parent nodes of the action are included. Therefore I need to get a pointer the Action node, which I do using lyd_find_xpath, and pass it to lyd_validate as the first argument. However, the siblings and predecessors are still there, so I need to add LYD_OPT_NOSIBLINGS into the list of validation options, otherwise this code:
```
3602     if (*node) {
3603         if (options & LYD_OPT_NOSIBLINGS) {
3604             /* ctx is NULL */
3605         } else {
3606             ctx = (*node)->schema->module->ctx;
3607 
3608             /* check that the node is the first sibling */
3609             while ((*node)->prev->next) {
3610                 *node = (*node)->prev;
3611             }
3612         }
3613     }
```
will move the `node` pointer to the left-most sibling of the Action node and then this condition fails once the code returns to Action during iteration:
```
3629             if (iter->parent && (iter->schema->nodetype & (LYS_ACTION | LYS_NOTIF))) {
3630                 if (!(options & LYD_OPT_ACT_NOTIF) || act_notif) {
3631                     LOGVAL(LYE_INELEM, LY_VLOG_LYD, iter, iter->schema->name);
3632                     LOGVAL(LYE_SPEC, LY_VLOG_LYD, iter, "Unexpected %s node \"%s\".",
3633                            (options & LYD_OPT_RPC ? "action" : "notification"), iter->schema->name);
3634                     goto cleanup;
3635                 }
3636                 act_notif = iter;
3637             }
3638 
```
If I add LYD_OPT_NOSIBLINGS, the validation then fails here:
```
3594             /* LYD_OPT_NOSIBLINGS cannot be set in this case */
3595             if (options & LYD_OPT_NOSIBLINGS) {
3596                 LOGERR(LY_EINVAL, "%s: invalid parameter (variable arg const struct lyd_node *data_tree with LYD_OPT_NOSIBLINGS).", __func__);
3597                 goto cleanup;
3598             }
```

Also I was wondering if validation or leafrefs inside action output with relative paths will work if do not include parents, but maybe yang 1.1 RFC doesn't allow this.
With the patch above libyang successfully detected unresolved leafref values returned in the action output.

Thanks,
Milan

**Comment by michalvasko** [comment]
Hi Milan,
you are right, creating action replies is not covered at all. I will make it possible using `lyd_new_output*` and `lyd_new_path()` with `LYD_PATH_OPT_OUTPUT` option. Will reference this issue once finished. If you then pass this tree into validation with some additional data tree, I believe it should work.

Regards,
Michal

**Comment by milanlenco** [comment]
Alright, I will comment out the unit test for action output for now and I will get back to it then. This issue is not a blocker for any of our current tasks.

Regards,
Milan

**Comment by michalvasko** [comment]
Problem solved in the current devel branch (0.11.0).

Regards,
Michal

**Comment by milanlenco** [comment]
Thanks, our unit tests also confirm that it is fixed now.
